Add user table and class

asaunier opened the pull request:
[Still a looot of work to do]

- [x] Create model for userdata => ~~Question: should we move the user-related tables (userdata, users_groups as in v5? etc.) out of schema "guidebook" to a dedicated "users" (?) DB schema?~~ => done
- [x] use v5 password hashed passwords

Related to #16

elemoine commented:
> Question: should we move the user-related tables (userdata, users_groups as in v5? etc.) out of schema "guidebook" to a dedicated "users" (?) DB schema?

A "user" DB schema would make sense to me.

elemoine commented:
Don't you want to store hashed passwords? As in c2cgeoportal. See https://github.com/camptocamp/c2cgeoportal/blob/master/c2cgeoportal/models.py#L185.

asaunier commented:
> Don't you want to store hashed passwords?

Sure. Passwords are already indeed hashed in v5.
Thanks for pointing what exists for c2cgeoportal.

tsauerwein commented:
See also: http://security.stackexchange.com/questions/17421/how-to-store-salt

asaunier commented:
I have moved the userdata table to schema "users" (I think the "user" table/document will stay in schema "guidebook").

I have also added the password handling functions from c2cgeoportal. And tests. The encryption used is from c2cgeoportal as well, but not sure we will keep it as is since we must be compatible with the encrypted passwords from the existing database (can we really convert the already encrypted passwords?). See https://github.com/c2corg/camptocamp.org/blob/master/lib/model/doctrine/UserPrivateData.class.php

asaunier commented:
Using HTTPS scheme for the API URLs is needed (highly recommended for OAuth and that's a wish from the client). I don't know if it is really possible for dev instances since waitress seems to not directly support https.

elemoine commented:
> Using HTTPS scheme for the API URLs is needed (highly recommended for OAuth and that's a wish from the client). I don't know if it is really possible for dev instances since waitress seems to not directly support https (unless by using a proxy in front of it and setting url_scheme = https in development.ini).

We may want to use mod_wsgi-express instead of waitress. With mod_wsgi-express the mod_wsgi author has done an excellent job on making it easier to use mod_wsgi.

gberaudo commented:
The PHP function used in v5 seems more secure than what is used in
c2cgeoportal. Why not keep the hashing logics?

On Fri, Sep 25, 2015 at 4:21 PM, Alex Saunier <notifications@github.com>
wrote:

> I have moved the userdata table to schema "users" (I think the "user"
> table/document will stay in schema "guidebook").
>
> I have also added the password handling functions from c2cgeoportal. And
> tests. The encryption used is from c2cgeoportal as well, but not sure we
> will keep it as is since we must be compatible with the encrypted passwords
> from the existing database (can we really convert the already encrypted
> passwords?). See
> https://github.com/c2corg/camptocamp.org/blob/master/lib/model/doctrine/UserPrivateData.class.php
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/c2corg/v6_api/pull/21#issuecomment-143236623>.
>

asaunier commented:
Anyway, we'll probably have to stick to the v5 encryption unless we re-encrypt the existing passwords (is it  even possible?).

tsauerwein commented:
> The PHP function used in v5 seems more secure than what is used in c2cgeoportal. Why not keep the hashing logics?

The [PHP function password_hash](http://php.net/manual/en/function.password-hash.php) uses bcrypt, which is [recommended](http://codahale.com/how-to-safely-store-a-password/). So I would also keep using bcrypt, for example with this [module](https://pypi.python.org/pypi/bcrypt/1.1.0).

> ...unless we re-encrypt the existing passwords (is it even possible?).

No, you can not get the plain-text passwords again. If we *would* want to use a different hashing method, we could do the following: Have two columns "old_password" (with the old password) and "password". Once a user logs in for the first time (at this moment we have the plain text password), we check with the column "old_password", calculate the new password hash and fill the column "password". For the next logins we will only check with the new password hash. 
But I am not saying we should do that, because bcrypt is a good option.

elemoine commented:
http://codahale.com/how-to-safely-store-a-password/ is a very interesting article. +1 on relying on bcrypt.

asaunier commented:
> So I would also keep using bcrypt, for example with this module.

Done at ff26661

gberaudo commented on c2corg_api/models/userdata.py:
```diff
@@ -0,0 +1,58 @@
+    __tablename__ = 'userdata'
+    __table_args__ = {"schema": users_schema}
+    __mapper_args__ = {
+        'polymorphic_identity': 'ud'
```
Do we need it to be polymorphic?

asaunier replied:
Not sure. But in c2cgeoportal we use it even for models with no inheritance:
https://github.com/camptocamp/c2cgeoportal/blob/master/c2cgeoportal/models.py#L196

gberaudo commented on c2corg_api/models/userdata.py:
```diff
@@ -0,0 +1,58 @@
+        'polymorphic_identity': 'ud'
+    }
+
+    document_id = Column(Integer, primary_key=True)
```
Is it a document?

asaunier replied:
No. It's different from the User documents.
But the user id should be related to the User document id.

By the way I'm a bit annoyed: that would make sense to rename this model to User but it would conflict with the future User document model. Or maybe we could name the user document as Profile for instance?

asaunier replied:
Perhaps we should rename this field to "id" and add a FK towards the User (or Profile?) document_id

tsauerwein replied:
> Perhaps we should rename this field to "id" and add a FK towards the User (or Profile?) document_id

I would make a field `id` as PK and a field `document_id` as FK pointing to the profile (which is a document).

gberaudo commented:
I renamed userdata to user and rebased on master.
I also changed the PK to `id`. We will add later the FK to the profile `document_id`.
I think we can merge it. Then in a subsequent PR I will add a view for creating a user.

tsauerwein commented on c2corg_api/models/user.py:
```diff
@@ -0,0 +1,57 @@
+    email = Column(String(200), nullable=False, unique=True)
+    email_validated = Column(Boolean, nullable=False)
+    _password = Column("password", String(255), nullable=False)
+    temp_password = Column(String(255))
```
What is `temp_password`?

asaunier replied:
It's used when the user asks for a new password ("forgot your password").

tsauerwein commented on c2corg_api/models/user.py:
```diff
@@ -0,0 +1,57 @@
+    username = Column(String(200), nullable=False, unique=True)
+    email = Column(String(200), nullable=False, unique=True)
+    email_validated = Column(Boolean, nullable=False)
+    _password = Column("password", String(255), nullable=False)
```
Why is there an underscore?

tsauerwein replied:
Sorry, I made the comment before seeing the property definition.

asaunier replied:
Because it is the encrypted version of the password. It is set using ``_set_password``:
https://github.com/c2corg/v6_api/blob/oauth/c2corg_api/models/user.py#L30-L31

tsauerwein commented on c2corg_api/tests/models/test_user.py:
```diff
@@ -0,0 +1,17 @@
+
+class Testuser(BaseTestCase):
+
+    def test_to_archive(self):
```
The name `test_to_archive` does not seem right.

asaunier replied:
Indeed. What about ``test_validate_password``?

tsauerwein commented on c2corg_api/models/user.py:
```diff
@@ -0,0 +1,57 @@
+        this method _MUST_ return a boolean.
+        @param passwd: the password that was provided by the user to
+        try and authenticate. This is the clear text version that we will
+        need to match against the (possibly) encrypted one in the database.
```
Why "possibly" encrypted?

asaunier replied:
I don't know, this code comes from GeoMapFish. I think we can remove "(possibly)".

tsauerwein commented:
Looks good to me.

asaunier commented:
Looks good to me too (except the test function name).

asaunier commented:
I have made the 2 minor changes mentioned previously (renamed test function and remove "possibly" in the validate password function comment).

I'm waiting for the Travis approval before merging the PR.

gberaudo commented:
Rebased, waiting for Travis.

asaunier commented:
Thanks, Travis passes. Merging...
